text/0000-numeric-literal-types.md

```diff
@@ -0,0 +1,318 @@
+explanation is written only with this full generality in mind, since I don't
+think it's too difficult to explain the weakenings. Accepting this RFC will
+probably entail picking a weakening and applying it to both explanations.
+
+# Motivation
+[motivation]: #motivation
+
+This proposal has a few motivating use cases:
+- Untyped compile-time constants, as in Go or C (via `#define`).
+- Bignum constructors.
+- Custom integer literals, à la `operator""`.
+
+The former is valuable, because it allows us to hoist several occurences
+of the same literal in different typed contexts, without having to type
+it as the largest possible numeric type and explicitly narrow, i.e.
+```rust
+let foo = my_u8() & 0b0101_0101;
+let bar = my_i32() & 0b0101_0101;
+// becomes
+const MY_MASK: ulit = 0b0101_0101;
+let foo = my_u8() & MY_MASK;
+let bar = my_i32() & MY_MASK;
+// instead of
+const MY_MASK: u128 = 0b0101_0101;
+let foo = my_u8() & (MY_MASK as u8);
+let bar = my_i32() & (MY_MASK as i32);
+```
+This can be emulated by a macro that expands to the given literal, but
+that is unergonomic, and calling `MY_MASK!()` does not make it clear
+that this is a compile-time constant (`ALL_CAPS` not withstanding).
+
+The latter two are essentially the same proposal: access to arbitrary-precission
+integers for constructing bignums (and other custom literals).
+Custom literals need to take a number as input; while
+C++, the only language with custom literals, simply takes its versions of
+`u64` and `f64` as arguments for literals, this is an unnecessary restriction
+in Rust, given that we recently stabalized the `u128` type. This problem
+cannot be neatly worked around, as far as we know.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Consider the expression `42`. What's its type? The basic language introduction
+might lead you to believe that, like in C++ and Java, it's `i32`. In reality,
+the compiler assigns this expression the type `ulit`: the type of all *integer
+literals*. Note the `u`: this is because all integer literals are unsigned!
+The float equivalent is `flit`.
+
+`ulit` and `flit` are both DSTs, so they can't be passed to functons like
+normal values. Unlike other DSTs, however, if they are used in a `Sized` 
+context, they will attempt to coerce into a sized integer type, defaulting
+to `i32` or `f64` if there isn't an obvious choice. This occurs silently,
+since one almost always wants a sized integer:
+```rust
+let x = 42; // 42 types as ulit, but since a let binding requires a 
+            // Sized value, it tries to coerce to a sized integer. since
+            // there isn't an obvious one, it picks i32. Hence, x: i32.
+
+let y = 42u32; // 42u32 has type u32, so no coersion occurs.
+let z: u32 = 42; // ulit coerces to u32, since it's the required type
+```
+
+Literal types are otherwise *mostly* like normal integers. They support
+arithmetic and comparisons (but don't implement any `std::ops` traits, since
```

Comment:
Does this include arithmetic for `flit`?

Reply:
Yes; I guess that's not clear. I'll update the RFC noting that later.

Reply:
*Why*? Where's the motivation for arbitrary-precision "float" arithmetic?

Also: *how*? Completely accurate arithmetic beyond integers is a difficult area without any local minima that are clearly good enough to bake into the language. Usually people envision bignum rationals for this sort of thing, but besides not being able to handle anything beyond add/sub/mul/div, it has serious downsides: even if you always use irreducible fractions and eat the significant overhead of doing so, the space requirements are pretty bad for many numbers (and truly awful for some).